capsules/src/usb/ctap.rs

```diff
@@ -0,0 +1,391 @@
+    /// this as a callback to mean that the transmission finished by waiting
+    /// until this function is called when we don't have anything left to send.
+    fn packet_in(&'a self, transfer_type: TransferType, endpoint: usize) -> hil::usb::InResult {
+        match transfer_type {
+            TransferType::Bulk => {
+                self.tx_buffer
+                    .take()
+                    .map_or(hil::usb::InResult::Delay, |tx_buf| {
+                        // Check if we have any bytes to send.
+                        let offset = self.tx_offset.get();
+                        let remaining = self.tx_len.get() - offset;
+                        if remaining > 0 {
+                            // We do, so we go ahead and send those.
+
+                            // Get packet that we have shared with the underlying
+                            // USB stack to copy the tx into.
+                            let packet = self.buffer(endpoint);
+
+                            // Calculate how much more we can send.
+                            let to_send = cmp::min(packet.len(), remaining);
+
+                            // Copy from the TX buffer to the outgoing USB packet.
+                            for i in 0..to_send {
+                                packet[i].set(tx_buf[offset + i]);
+                            }
+
+                            // Update our state on how much more there is to send.
+                            self.tx_offset.set(offset + to_send);
+
+                            // Put the TX buffer back so we can keep sending from it.
+                            self.tx_buffer.replace(tx_buf);
+
+                            // Return that we have data to send.
+                            hil::usb::InResult::Packet(to_send)
+                        } else {
+                            // We don't have anything to send, so that means we are
+                            // ok to signal the callback.
+
+                            // Signal the callback and pass back the TX buffer.
+                            self.tx_client.map(move |tx_client| {
+                                tx_client.transmitted_buffer(
+                                    tx_buf,
+                                    self.tx_len.get(),
+                                    ReturnCode::SUCCESS,
+                                )
+                            });
+
+                            // Return that we have nothing else to do to the USB
+                            // driver.
+                            hil::usb::InResult::Delay
+                        }
+                    })
+            }
+            TransferType::Control | TransferType::Isochronous | TransferType::Interrupt => {
+                unimplemented!();
+            }
+        }
+    }
+
+    /// Handle a Bulk/Interrupt OUT transaction
+    fn packet_out(
+        &'a self,
+        transfer_type: TransferType,
+        endpoint: usize,
```

Comment:
Can endpoints be an enum? Endpoint 0xffffffff seems unlikely.

Reply:
That is a large change. This function comes from the HIL so the USB HIL would need to be changed as well as everything that uses the HIL. Then the USB packet decoded to a enum (not that hard).

What advantage would that get us?

Reply:
I'm going to mark this as resolved. As part of the general USB TODO (https://github.com/tock/tock/issues/1940) we want to convert USB endpoints to be enums. That needs to happen in the HIL though.